packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts

```diff
@@ -120,16 +134,43 @@ class TlsValidationFileTrust extends TlsValidationTrust {
 
   public bind(_scope: Construct): TlsValidationTrustConfig {
     return {
-      virtualNodeClientTlsValidationTrust: {
+      tlsValidationTrust: {
         file: {
           certificateChain: this.certificateChain,
         },
       },
-      virtualGatewayClientTlsValidationTrust: {
-        file: {
-          certificateChain: this.certificateChain,
+    };
+  }
+}
+
+class TlsValidationSdsTrust extends TlsValidationTrust {
+  /**
+   * The name of the secret for Envoy to fetch from a specific endpoint through the Secrets Discovery Protocol.
+   */
+  readonly secretName: string;
+
+  constructor (props: TlsValidationSdsTrustOptions) {
+    super();
+    this.secretName = props.secretName;
+  }
+
+  public bind(_scope: Construct): TlsValidationTrustConfig {
+    return {
+      tlsValidationTrust: {
+        sds: {
+          secretName: this.secretName,
         },
       },
     };
   }
 }
+
+/**
+ * Represents the properties needed to define subject alternative names
+ */
+export interface SubjectiveAlternativeNames {
```

Comment:
I have to say, this is a really weird API...

What's the reason for modeling it this way? Is it "future proofing", because you know there will be other ways to match on SANs in the future?

Reply:
Thanks for pointing this out. We actually had some discussion on how to model this API.

First and foremost, you are correct about the "future proofing". We are quite certain there will be more match patterns be added. In addition, most likely `SubjectAlternativeNames` allows defining multiple match patterns.

For above two reasons, we modeled `SubjectAlternativeNames` as a interface with single property for now.

---
P.S:  I will include this in the commit message.

Reply:
Yes, I have an idea. Let's model this similarly to how we modeled [`HttpHeaderMatch`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-appmesh.HttpHeaderMatch.html):

```ts
export interface TlsValidation {
  // ...

  readonly subjectAlternativeNames?: SansMatch;
}

abstract class SansMatch {
  public static exactly(...sans: string[]): SansMatch {
    // ....
```

(Names are TBD, of course, feel free to improve on them)

Reply:
Actually our original plan was to use this method but we decided not to since **it's possible we'll allow users to define multiple match modes.** 

`HttpHeaderMatch` model works because [`headers`](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-appmesh/lib/route-spec.ts#L45) property is an array.

Please refer this comment for [more reference](https://github.com/Seiya6329/aws-cdk/pull/3#discussion_r647642281).

Reply:
I'm not worried about that at all - in the future, if you add support for multiple match modes, we can have an API like:

```
  subjectAlternativeNames: appmesh.SansMatch.or(SansMatch.exactly('abc', 'def'), SansMatch.startWith('xyz')),
```

Which I like even more than the implicit "or" that an array suggests.

Reply:
Thanks for providing further detail.

**I have one question:** do you expect to write `.or()` method to prohibit accepting same type of factory method? I am thinking in order to prohibit it, factory method might need to be configured to return different Object.

Reply:
Yes. I would probably employ a trick very similar to what I described for `TlsCertificate` in this PR:

```ts
export class SansMatch {
  public static exactly(...names: string[]): SimpleSansMatch;

  public static or(...matchers: SimpleSansMatch[]): SansMatch;

  // ...
}

export class SimpleSansMatch extends SansMatch {
}
```

Reply:
I agree with @skinny85. We should use the same pattern for this API that we are for route matches. The `or` method is also pretty interesting. Should we be using that for the header match?

Reply:
Thanks for @skinny85 and @dfezzie for looking close into this subject.

I am making the change as suggested here for two clear benefits:

- Be consistent of using factory method
- Using `.or()` makes more explicit that multiple match patterns are selected